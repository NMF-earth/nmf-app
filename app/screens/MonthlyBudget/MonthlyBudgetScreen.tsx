import React, { useState } from "react";
import { View, Slider, ScrollView } from "react-native";
import { Ionicons } from "@expo/vector-icons";
import * as WebBrowser from "expo-web-browser";
import { useDispatch, useSelector } from "react-redux";

import { Text, Button } from "../../components";
import { Colors } from "../../style";
import styles from "./MonthlyBudgetScreen.styles";
import navigationOptions from "./MonthlyBudgetScreen.navigationOptions";
import { t } from "../../utils";
import { budget } from "../../ducks";
import { navigate } from "../../navigation";

const MIN_MONTHLY_CARBON_BUDGET = 0;
const MAX_MONTHLY_CARBON_BUDGET = 1000;

const translationMontlyBudgetCountries = [
  "MONTHLY_BUDGET_LUXEMBOURG",
  "MONTHLY_BUDGET_UNITED_STATES",
  "MONTHLY_BUDGET_JAPAN",
  "MONTHLY_BUDGET_SWEDEN",
  "MONTHLY_BUDGET_FRANCE",
  "MONTHLY_BUDGET_CHINA",
  "MONTHLY_BUDGET_BRAZIL",
  "MONTHLY_BUDGET_INDIA",
  "MONTHLY_BUDGET_ETHIOPIA"
];

const CountryExample = (translation, index) => (
  <Text.Secondary center key={index} style={styles.worldExampleItem}>
    {t(translation)}
  </Text.Secondary>
);

const MonthlyBudgetScreen = ({ navigation }) => {
  const monthlyBudget = useSelector(budget.selectors.getMonthlyCarbonBudget);
  const navigator = navigate(navigation);

  const [sliderValue, setSliderValue] = useState(monthlyBudget);
  const dispatch = useDispatch();

  const onPressInfo = () =>
    WebBrowser.openBrowserAsync(
      "https://en.wikipedia.org/wiki/List_of_countries_by_carbon_dioxide_emissions_per_capita"
    );

  const onPressSaveBudget = () => {
    dispatch(budget.actions.setMonthlyCarbonBudget(Math.round(sliderValue)));
    navigator.goBack();
  };

  return (
    <View style={styles.container}>
      <ScrollView>
        <View style={styles.personnalBudgetContainer}>
          <Text.Primary bold>{t("MONTHLY_BUDGET_SLIDE_TO_SET")}</Text.Primary>
        </View>
        <Slider
          minimumTrackTintColor={Colors.linkGreen}
          maximumTrackTintColor={Colors.gray}
          thumbTintColor={Colors.linkGreen}
          style={styles.slider}
          maximumValue={MAX_MONTHLY_CARBON_BUDGET}
          minimumValue={MIN_MONTHLY_CARBON_BUDGET}
          value={sliderValue}
          onSlidingComplete={setSliderValue}
        />
        <Text.Primary lightGray>
          {Math.round(sliderValue) + " kg CO2eq"}
        </Text.Primary>
        <View style={styles.worldBudgetContainer}>
          <View style={styles.worldExampleTitle}>
            <Text.Primary bold>
              {t("MONTHLY_BUDGET_CARBON_EMISSIONS_WORLD")}
              <Ionicons
                name="md-information-circle"
                size={26}
                color={Colors.linkGreen}
                onPress={onPressInfo}
              />
            </Text.Primary>
          </View>
          {translationMontlyBudgetCountries.map(CountryExample)}
          <View style={styles.parisAgreement}>
            <Text.Secondary center>
              {t("MONTHLY_BUDGET_PARIS_AGREEMENT")}
              <Text.Secondary bold green>
                {" 166 kg CO2"}
              </Text.Secondary>
            </Text.Secondary>
          </View>
        </View>
      </ScrollView>
      <View style={styles.buttonContainer}>
        <Button.Primary onPress={onPressSaveBudget} textType={"Primary"}>
          <Text.Primary white center bold>
            {t("MONTHLY_BUDGET_SAVE")}
          </Text.Primary>
        </Button.Primary>
      </View>
    </View>
  );
};

MonthlyBudgetScreen.navigationOptions = navigationOptions;

export default MonthlyBudgetScreen;
