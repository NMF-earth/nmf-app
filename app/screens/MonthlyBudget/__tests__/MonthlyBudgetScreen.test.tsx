import React from "react";
import { create } from "react-test-renderer";

import MonthlyBudgetScreen from "../MonthlyBudgetScreen";

const props = {
  navigation: {
    goBack: jest.fn(),
  },
};

it("MonthlyBudgetScreen renders correctly", () => {
  const tree = create(<MonthlyBudgetScreen {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("MonthlyBudgetScreen should go back if save button is pressed", () => {
  const root = create(<MonthlyBudgetScreen {...props} />).root;
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore
  const button = root.findByType("Button.Primary");

  button.props.onPress();
  expect(props.navigation.goBack).toHaveBeenCalled();
});
