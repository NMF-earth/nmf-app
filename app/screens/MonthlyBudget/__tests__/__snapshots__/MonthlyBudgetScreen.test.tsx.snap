// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthlyBudgetScreen renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "paddingHorizontal": 16,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "paddingTop": 26,
          }
        }
      >
        <Text.Primary
          bold={true}
        >
          Slide to set your budget
        </Text.Primary>
      </View>
      <RNCSlider
        disabled={false}
        enabled={true}
        inverted={false}
        maximumTrackTintColor="#CED4DA"
        maximumValue={1000}
        minimumTrackTintColor="#4BB377"
        minimumValue={10}
        onChange={null}
        onRNCSliderAccessibilityAction={null}
        onRNCSliderSlidingComplete={[Function]}
        onRNCSliderSlidingStart={null}
        onRNCSliderValueChange={null}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        step={0}
        style={
          Array [
            Object {
              "height": 40,
            },
            Object {
              "marginTop": 14,
            },
          ]
        }
        tapToSeek={false}
        thumbImage={null}
        thumbTintColor="#4BB377"
        value={0}
      />
      <Text.Primary
        lightGray={true}
      >
        NaN kg CO2eq
      </Text.Primary>
      <View
        style={
          Object {
            "flex": 1,
            "marginVertical": 20,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "marginBottom": 16,
            }
          }
        >
          <Text.Primary
            bold={true}
          >
            Average monthly emissions per capita in a sample of countries 
            <Ionicons
              color="#3957D0"
              name="md-information-circle"
              onPress={[Function]}
              size={26}
            />
          </Text.Primary>
        </View>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          Luxembourg : 3.5 tons
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          United States : 1.5 tons
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          Japan : 900 kg
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          Sweden : 595 kg
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          France : 575 kg
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          China : 522 kg
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          Brazil : 208 kg
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          India : 139 kg
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          Ethiopia : 8.3 kg
        </Text.Secondary>
        <View
          style={
            Object {
              "marginBottom": 80,
              "marginTop": 16,
            }
          }
        >
          <Text.Secondary
            center={true}
          >
            If you wish to respect the Paris agreement (to keep the rise in global average temperature below 2 degrees), set your monthly budget at 167 kgCO2eq
            <Ionicons
              color="#3957D0"
              name="md-information-circle"
              onPress={[Function]}
              size={26}
            />
          </Text.Secondary>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "bottom": 20,
        "left": 16,
        "position": "absolute",
        "right": 16,
      }
    }
  >
    <Button.Secondary
      fullWidth={true}
      icon="save"
      onPress={[Function]}
      text="Save"
    />
  </View>
</View>
`;
